src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,411 @@
+        tag.setInt(DIMENSION, snapshot.world.getDimension().getType().getId());
+        snapshot.region.serializeTo(tag);
+
+        // Palette serialization begin
+        NBTTagList palettes = new NBTTagList();
+        Object2IntMap<IBlockState> mapPalettes = new Object2IntOpenHashMap<>();
+        Set<IBlockState> recordedPalettes = new HashSet<>();
+
+        // "nothing should be here" block state
+        palettes.add(new NBTTagCompound());
+        // Palette serialization end
+
+        // Streaking: a way to record a number of repeating block states, using only 2 integers
+        // Storage frame: one or two numbers indicating a sequence of block states
+        // Regular frame: a type of storage frame; 1 number, representing the ID of the block state recorded
+        // Streak frame: a type of storage frame; 2 numbers, where the first one is positive, representing the ID of the repeating block state; the second one is negative, representing the number of repeating block states
+
+        // Complete storage frames, will be serialized into a NBTTagList later
+        IntList frames = new IntArrayList();
+
+        // The repeating block state, if any
+        IBlockState lastStreak = null;
+        // Number of repeating block states
+        int streak = 0;
+
+        boolean firstIteration = true;
+        for (IBlockState state : snapshot.blockStates) {
+            // Palette serialization begin
+            if (!recordedPalettes.contains(state)) {
+                recordedPalettes.add(state);
+                // ID 0 is preoccupied for "nothing should be placed here", which is represented in 'null'
+                mapPalettes.put(state, mapPalettes.size() + 1);
+                palettes.add(NBTUtil.writeBlockState(state));
+            }
+            // Palette serialization ends
+
+            // Hence we record the storage frame after it is completed, the inital values of the counter would be used as a storage frame regardless, therefore we need to skip inorder for this algorithm to function correctly
+            if (firstIteration)
+                firstIteration = false;
+            else {
+                // If the same block state showed up again, increase the number of streaks
+                // If we have more than 255 streaks (maximum integer that can be stored in 8 bits), we force start a new streak, because the current storage frame won't fit that many of them
+                if (state == lastStreak && streak < 255) {
+                    streak++;
+                    continue;
+                }
+
+                // Note: if the program reached here, it means the current storage frame is completed
+
+                // If it didn't show up and this block is not by itself, record the streak frame
+                // The first one represents the ID of the streaking block state, the second one, which is negative, represents the amount of streaking block states
+                if (streak > 1) {
```

Comment:
This needs a check whether we overflowed the one byte we've got available and if so start a new streak...
Very unlikely, but theoretically possible...